### PR TITLE
[Hexagon] Remove TVM logging from performance-sensitive parts of runtime

### PR DESCRIPTION
**Pull request by kparzysz-quic**
Use HAP messaging instead. This greatly reduced execution overhead.

### CODE REVIEW

**Comment by kparzysz-quic** [comment]
Please don't use logging.h (LOG, CHECK, etc.) in the Hexagon RT, at least in the performance-sensitive parts of it.
@janetsc @supersat @adstraw

**Comment by kparzysz-quic** [comment]
A large part of the slowness came from exception handling, specifically from the failing assertion in `FreeHexagonBuffer` about trying to free an unknown buffer.  There is something weird going on there and I have disabled that assertion (left the message).

File: src/runtime/hexagon/hexagon_buffer_manager.h [hunk 1]
```diff
@@ -39,11 +38,17 @@ class HexagonBufferManager {
...
+      // This should be an assertion, but something seems to go wrong here.
+      // The symptom is that when resources are being released (ReleaseResources),
+      // one buffer disappears from the "runtime" buffer manager, and is not deleted
+      // when that manager is reset. The FreeHexagonBuffer is than called for that
+      // buffer, but with the "static" buffer manager instead. That manager doesn't
+      // find it in the map and throws an exception, which somehow doesn't abort
+      // the program.
+      HEXAGON_PRINT(ERROR, "Attempt made to free unknown or already freed dataspace allocation");
+    } else {
+      HEXAGON_ASSERT(it->second != nullptr);
```
Comment (janetsc) [inline src/runtime/hexagon/hexagon_buffer_manager.h:51]: I will investigate the buffer free after ReleaseResources.  This happens on session shutdown when the process is already being torn down.

**Review by janetsc — COMMENTED** [review]
Can you quantify the savings you saw with this change?

**Comment by kparzysz-quic** [comment]
Our local simulator tests would time out after 1h, now they finish within 30 min.  I used a subset of the fp16 conv tests, and without this change they would take about 45s per test, after the change they take about 15s per test.

**Comment by janetsc** [comment]
> 



> Our local simulator tests would time out after 1h, now they finish within 30 min. I used a subset of the fp16 conv tests, and without this change they would take about 45s per test, after the change they take about 15s per test.

Have your simulator runs always taken this long, or did this recently start happening?  I am wondering if I introduced this will resource management, and this points to needing to handle Free better after ReleaseResources has been called.  (And that's the root culprit, not the logging, which should only be processed upon a failed check.)

**Comment by kparzysz-quic** [comment]
The degradation started a while back, but we didn't notice, because the tests were passing without timing out.  The main contributor to the degradation is the exception being thrown, but I only narrowed it down after creating this PR (I used it locally to unblock our tests---that was an urgent thing to do).

I can revert it, and only remove the actual assertion in free-after-release.  That situation shouldn't be happening---there is something weird going on there.

**Comment by janetsc** [comment]
> The degradation started a while back, but we didn't notice, because the tests were passing without timing out. The main contributor to the degradation is the exception being thrown, but I only narrowed it down after creating this PR (I used it locally to unblock our tests---that was an urgent thing to do).
> 
> I can revert it, and only remove the actual assertion in free-after-release. That situation shouldn't be happening---there is something weird going on there.

I tested this locally, which I prefer to removing the assert.  This will catch the error if we are operating on the static buffer manager:

```
void HexagonDeviceAPI::FreeDataSpace(Device dev, void* ptr) {
  CHECK(ptr) << "buffer pointer is null";
  CHECK(IsValidDevice(dev)) << "dev.device_type: " << dev.device_type;
  if (mgr == &hexbuffs) {
    // Free can throw if it is called after ReleaseResources
    try {
      mgr->FreeHexagonBuffer(ptr);
    } catch (int e) {
      LOG(INFO) << "FreeDataSpace called for static manager.  An exception occurred " << e;
    }
  } else {
    mgr->FreeHexagonBuffer(ptr);
  }
}
```

**Comment by kparzysz-quic** [comment]
> I tested this locally, which I prefer to removing the assert in `FreeHexagonBuffer`. This will catch the error if we are operating on the static buffer manager:

This still throws an exception.  The whole point of this is not to throw it.

**Comment by janetsc** [comment]
> 

Ah, I see some tests are throwing this even for runtime_hexbuffs (before ReleaseResources is called.). I'm going to continue to investigate.

**Comment by kparzysz-quic** [comment]
Closing this, will come up with a smaller fix.

**Pull request closed without merging**